File: misc/matrix_rotation/matrix_rotation.java
import java.io.*;
import java.util.*;

public class matrix_rotation {

	public static void print(int[][] matrix) {
		int M = matrix.length;
		int N = matrix[0].length;
		for(int r = 0; r < M; ++r) {
			for(int c = 0; c < N; ++c)
				System.out.printf("%3d", matrix[r][c]);
			System.out.println();
		}
	}

	public static void rotate_horizontal(int[][] matrix) {
		int N = matrix.length;
		for(int r = 0; r < N; ++r) {
			for(int c = 0; c < matrix[r].length / 2; ++c) {
				int temp = matrix[r][c];
				matrix[r][c] = matrix[r][N-1-c];
				matrix[r][N-1-c] = temp;
			}
		}
	}

	public static void rotate_vertical(int[][] matrix) {
		int N = matrix.length;
		for(int r = 0; r < N / 2; ++r) {
			for(int c = 0; c < matrix[r].length; ++c) {
				int temp = matrix[r][c];
				matrix[r][c] = matrix[N-1-r][c];
				matrix[N-1-r][c] = temp;
			}
		}
	}

	public static void rotate_1st_diagonal(int[][] matrix) {
		int N = matrix.length;
		for(int r = 0; r < N; ++r) {
			for(int c = r; c < N; ++c) {
				int temp = matrix[r][c];
				matrix[r][c] = matrix[c][r];
				matrix[c][r] = temp;
			}
		}
	}

	public static void rotate_2nd_diagonal(int[][] matrix) {
		int N = matrix.length;
		for(int r = 0; r < N; ++r) {
			for(int c = 0; c < N-r; ++c) {
				int temp = matrix[r][c];
				matrix[r][c] = matrix[N-1-c][N-1-r];
				matrix[N-1-c][N-1-r] = temp;
			}
		}
	}

	public static void spin_90degree_clockwise(int[][] matrix) {
		rotate_1st_diagonal(matrix);
		rotate_horizontal(matrix);
	}

	public static void main(String[] args) {

		int[][] m1 = new int[][] {
			{ 1, 2, 3 },
			{ 4, 5, 6 },
			{ 7, 8, 9 },
		};
		System.out.println("Before rotate horizontally:");
		print(m1);
		rotate_horizontal(m1);
		System.out.println("After  rotate horizontally:");
		print(m1);

		int[][] m2 = new int[][] {
			{ 1, 2, 3 },
			{ 4, 5, 6 },
			{ 7, 8, 9 },
		};
		System.out.println("Before rotate vertically:");
		print(m2);
		rotate_vertical(m2);
		System.out.println("After  rotate vertically:");
		print(m2);

		int[][] m3 = new int[][] {
			{ 1, 2, 3 },
			{ 4, 5, 6 },
			{ 7, 8, 9 },
		};
		System.out.println("Before rotate diagonal:");
		print(m3);
		rotate_1st_diagonal(m3);
		System.out.println("After  rotate diagonal:");
		print(m3);

		int[][] m4 = new int[][] {
			{ 1, 2, 3 },
			{ 4, 5, 6 },
			{ 7, 8, 9 },
		};
		System.out.println("Before rotate 2nd diagonal:");
		print(m4);
		rotate_2nd_diagonal(m4);
		System.out.println("After  rotate 2nd diagonal:");
		print(m4);

		int[][] m5 = new int[][] {
			{ 1, 2, 3 },
			{ 4, 5, 6 },
			{ 7, 8, 9 },
		};
		System.out.println("Before spinning 90 degree clockwise:");
		print(m5);
		spin_90degree_clockwise(m5);
		System.out.println("After  spinning 90 degree clockwise:");
		print(m5);
	}
}

File: misc/matrix_rotation/matrix_rotation.py
#!/usr/bin/python3
import sys

# Print MxN matrix row by col
def print_matrix(matrix):
	M = len(matrix)
	N = len(matrix[0])
	if M == 0 or N == 0:
		return;
	for r in range(M):
		for c in range(N):
			print("{0:3}".format(matrix[r][c]),end='')
		print()

# Rotate horizontally
def rotate_horizontal(matrix):
	M = len(matrix)
	N = len(matrix[0])
	for r in range(M):
		for c in range(N // 2):
			temp = matrix[r][c]
			matrix[r][c] = matrix[r][N-1-c]
			matrix[r][N-1-c] = temp

# Rotate vertically
def rotate_vertical(matrix):
	M = len(matrix)
	N = len(matrix[0])
	for r in range(M // 2):
		for c in range(N):
			temp = matrix[r][c]
			matrix[r][c] = matrix[N-1-r][c]
			matrix[N-1-r][c] = temp

# Rotate by first diagonal
def rotate_1st_diagonal(matrix):
	N = len(matrix)
	for r in range(N):
		for c in range(r+1,N):
			if(r == c):
				continue
			temp = matrix[r][c]
			matrix[r][c] = matrix[c][r]
			matrix[c][r] = temp

# Rotate by second diagonal
def rotate_2nd_diagonal(matrix):
	N = len(matrix)
	for r in range(N):
		for c in range(N-1-r):
			if(r == N-1-c):
				continue
			temp = matrix[r][c]
			matrix[r][c] = matrix[N-1-c][N-1-r]
			matrix[N-1-c][N-1-r] = temp

def spin_90degree_clockwise(matrix):
	rotate_2nd_diagonal(matrix)
	rotate_vertical(matrix)

if __name__ == "__main__":

	# Rotate horizontally
	m1 = [
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	]
	print("Before rotate horizontally:")
	print_matrix(m1)
	rotate_horizontal(m1)
	print("After  rotate horizontally:")
	print_matrix(m1)
	print()

	# Rotate vertically
	m2 = [
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	]
	print("Before rotate vertically:")
	print_matrix(m2)
	rotate_vertical(m2)
	print("After  rotate vertically:")
	print_matrix(m2)
	print()

	# Rotate by first diagonal
	m3 = [
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	]
	print("Before rotate 1st diagonal:")
	print_matrix(m3)
	rotate_1st_diagonal(m3)
	print("After  rotate 1st diagonal:")
	print_matrix(m3)
	print()

	# Rotate by second diagonal
	m4 = [
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	]
	print("Before rotate 2nd diagonal:")
	print_matrix(m4)
	rotate_2nd_diagonal(m4)
	print("After  rotate 2nd diagonal:")
	print_matrix(m4)
	print()

	# Spin 90 degree clockwise
	m5 = [
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	]
	print("Before spinning 90 degree clockwise:")
	print_matrix(m5)
	spin_90degree_clockwise(m5)
	print("After  spinning 90 degree clockwise:")
	print_matrix(m5)
	print()

	print("Done")
